pkg/extension/controller/controller.go

```diff
@@ -118,17 +107,66 @@ func (ec *ExtensionController) Subscribe(ctx context.Context, reconcileChan chan
 }
 
 func (ec *ExtensionController) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	eventType, exists := ec.eventCache.popEvent(request.Namespace, request.Name)
+	if !exists {
+		eventType = EventTypeUnknown
+	}
+
+	ctx = ec.setupContext(ctx)
+
+	// Ensure finalizer exists for both create and updates
+	if eventType == EventTypeCreate || eventType == EventTypeUpdate {
+		if err := ec.ensureFinalizer(ctx, request); err != nil {
+			return reconcile.Result{}, err
+		}
+	}
+
+	// Call user reconcile function
+	ec.logger.Info("reconciling request", "namespace", request.Namespace, "name", request.Name, "event", eventType)
+	result, err := ec.reconcile(ctx, request, ec)
+	if err != nil {
+		return result, err
+	}
+
+	if eventType == EventTypeUpdate {
+		if err := ec.cleanupFinalizer(ctx, request); err != nil {
+			return reconcile.Result{}, err
+		}
```

Comment:
I will monitor how this goes, but there may be a need for us to Requeue this reconcile and re-insert the `UPDATE` into the eventCache on error case